Closes #71: Use results file instead of API for lcps

Miraeld opened the pull request:
# Description

Fixes #71 

## Documentation

### User documentation

### Technical documentation

This PR changes the way we are testing the result from the beacon script. In fact, before we were using Google PSI API, while now we are using a file to compare our results to the expected one. 
This allow us to faster, more reliable. 

## Type of change
- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies

## Risks

# Checklists

## Feature validation

- [ ] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [ ] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Documentation

- [ ] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [ ] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [ ] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [ ] I wrote self-explanatory code about what it does.
- [ ] I wrote comments to explain why it does it.
- [ ] I named variables and functions explicitely.
- [ ] I protected entry points against unexpected inputs.
- [ ] I did not introduce unecessary complexity.
- [ ] I listed the introduced external dependencies explicitely on the PR.
- [ ] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [ ] I handled errors when needed.
- [ ] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [ ] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [ ] I explicitely mentioned performance risks in the PR.
- [ ] I explicitely mentioned security risks in the PR.

Miraeld committed:
Closes #71: Use results file instead of API for lcps

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -26,49 +27,55 @@ const [actual, expected]: [LcpData, LcpData] = [{}, {}];
 
         // Get the LCP/ATF from the DB
-        sql = `SELECT lcp, viewport FROM ${tablePrefix}wpr_above_the_fold WHERE url LIKE "%${row[0]}%"`;
-        result = await dbQuery(sql);
```
Why did we delete this line

Miraeld replied:
The SQL wasn't deleted but updated and reformatted to be easier to read. 
About the `result` variable, I forgot to put it back. It's been modified now.

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -26,49 +27,55 @@ const [actual, expected]: [LcpData, LcpData] = [{}, {}];
     });
 
-    data = dataTable.rows();
+    const data = await fs.readFile(resultFile, 'utf8');
```
Why did this become a local constant? We should keep this global so the `Then` step definition can share the state.

Miraeld replied:
I've changed it.

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
There are some lint errors for this file:
- We can remove the axios package as we don't use the PSI api anymore.
- The `LcpDataTable[]` type is not longer needed and we can remove it from imported modules and also remove the type definition [here](https://github.com/wp-media/wp-rocket-e2e/blob/3223b038a7706ae1ade2ffba58bf14449a53f1b1/utils/types.ts#L103-L106)

Miraeld replied:
Okay, I've done both of feedback.

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -77,74 +84,41 @@ Given('I visit the following urls in {string}', async function (this: ICustomWor
+    let expectedResults: any;
+    // Read the expected results from the file
+    try {
+        const data = await fs.readFile(`./src/support/results/expectedResults${formFactor.charAt(0).toUpperCase() + formFactor.slice(1)}.json`, 'utf8');
```
We would not need this if the state of data is shared.

Miraeld replied:
Yep, it's been updated.

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -77,74 +84,41 @@ Given('I visit the following urls in {string}', async function (this: ICustomWor
-        } catch (error) {
-            console.error(`Error fetching PageSpeed Insight for ${url}:`, error);
-        }
+    let expectedResults: any;
```
We need to have a proper type definition here, using any can be problematic down the line.

Miraeld replied:
It's been deleted as we share the `data` variable now.

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -77,74 +84,41 @@ Given('I visit the following urls in {string}', async function (this: ICustomWor
-
-            console.log(`LCP for ${url} is ${lcp}`);
-            // Populate the expected data.
-            expected[row[0]] = {
```
Why was this removed?

Miraeld replied:
I did remove it indeed because we were looking at the API response.

Miraeld committed:
Merge branch 'trunk' into enhancement/71-use-result-file-for-lcp-tests

Miraeld committed:
Comply feedback

jeawhanlee committed:
Merge branch 'trunk' of https://github.com/wp-media/wp-rocket-e2e into trunk

jeawhanlee committed:
Removed unused hook

Miraeld committed:
Fix bugs + Update tests results

Miraeld commented:
Thia last commit corrects the expected results for Desktop. 
Everything pass for the desktop.

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -11,143 +11,106 @@
+import fs from 'fs/promises';
 
-let data: LcpDataTable[],
+let data: string,
     truthy: boolean = true,
     failMsg: string = "";
```
```suggestion
let data: string,
    truthy: boolean = true,
    failMsg: string = "",
    jsonData: Record<string, { lcp: string; viewport: string[] }>;
```

Miraeld replied:
Done

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -11,143 +11,106 @@
 
-    data = dataTable.rows();
+    data = await fs.readFile(resultFile, 'utf8');
+    const jsonData = JSON.parse(data);
```
```suggestion
    jsonData = JSON.parse(data);
```

Miraeld replied:
Done

jeawhanlee commented on src/support/steps/lcp-beacon-script.ts:
```diff
@@ -11,143 +11,106 @@
-                }
-            }
+Then('lcp and atf should be as expected', async function (this: ICustomWorld) {
+    const jsonData = JSON.parse(data);
```
```suggestion
```

Miraeld replied:
Done

Miraeld committed:
Updates expected results
Add is_mobile is sql query
Accept array for lcps

Miraeld committed:
Comply linter

Miraeld committed:
Comply feedback

Miraeld commented:
Okay so, 
@jeawhanlee I've updated with the last few commits the following: 

- Updated the JSON files based on QA feedbacks. 
- Introduced a `enabled` attribute for each entry of the JSON (as we can't comment out in a JSON). This attribute will allow us to enable or disable a URL in the test. 
- I've modified the LCP structure to be able to get an array as sometime we get several lcps for an img set. 

Everything is working and passing, all green on e2e. :D

Mai-Saad commented on src/support/results/expectedResultsMobile.json:
```diff
@@ -0,0 +1,236 @@
+    ],
+    "enabled": true
+  },
+  "lcp_single_double": {
```
What shall be the value here?

Mai-Saad replied:
if we don't handle this, then we can add the expected value and flag it as false so won't be executed for now. what do you think? https://wp-media.slack.com/archives/C070482U1TK/p1715763660411819

Mai-Saad commented on src/support/results/expectedResultsDesktop.json:
```diff
@@ -0,0 +1,234 @@
+    ],
+    "enabled": true
+  },
+  "lcp_single_double": {
```
What shall be the value here?

Mai-Saad reviewed (commented):
"lcp_single_double": has no value in json. @Miraeld can you please check

jeawhanlee committed:
Updated fixture

Mai-Saad commented:
@jeawhanlee thanks for the update. 2 tests failed and many has no data in database => can you please check? (I think we need to see if it's the same before the PR and while using API)
![Screenshot from 2024-05-15 17-28-51](https://github.com/wp-media/wp-rocket-e2e/assets/76941962/7912c9a8-442d-4b5d-a582-d4e989c0c754)
![Screenshot from 2024-05-15 17-31-11](https://github.com/wp-media/wp-rocket-e2e/assets/76941962/5c39782e-38a7-4c8f-971d-6a0d562499e0)

jeawhanlee commented:
@Mai-Saad I think I found a reason for the no result in db, since you are on e2e env, your `WP_SSH_ROOT_DIR` should end with `/htdocs`

jeawhanlee committed:
Updated hardcoded atf urls to relative (#81)

MathieuLamiot commented:
Here are my results. I do get the 2 failing tests, but I don't have the missing data in the DB.

mathieu@air-de-mathieu wp-rocket-e2e % npm run test:lcp

```
> wp-rocket-e2e@1.0.0 test:lcp
> $npm_package_config_testCommand --tags @lcp

This Node.js version (v21.6.2) has not been tested with this version of Cucumber; it should work normally, but please raise an issue if you see anything unexpected.
Error deleting folder "./backstop_data/bitmaps_test": ENOENT: no such file or directory, lstat './backstop_data/bitmaps_test'
Loading config:  /Users/mathieu/Documents/Github/wp-rocket-e2e/backstop.json 

COMMAND | Executing core for "reference"
  clean | backstop_data/bitmaps_reference was cleaned.
createBitmaps | Selected 6 of 6 scenarios.
CREATING NEW REFERENCE FILE
CREATING NEW REFERENCE FILE
CREATING NEW REFERENCE FILE
CREATING NEW REFERENCE FILE
CREATING NEW REFERENCE FILE
Browser Console Log 0: JSHandle:JQMIGRATE: Migrate is installed with logging active, version 3.4.1
Browser Console Log 0: JSHandle:JQMIGRATE: Migrate is installed with logging active, version 3.4.1
Browser Console Log 0: JSHandle:JQMIGRATE: Migrate is installed with logging active, version 3.4.1
Browser Console Log 0: JSHandle:JQMIGRATE: Migrate is installed with logging active, version 3.4.1
Browser Console Log 0: JSHandle:BackstopTools have been installed.
Browser Console Log 0: JSHandle:BackstopTools have been installed.
Browser Console Log 0: JSHandle:BackstopTools have been installed.
Browser Console Log 0: JSHandle:BackstopTools have been installed.
Browser Console Log 0: JSHandle:BackstopTools have been installed.
x Close Browser
x Close Browser
x Close Browser
x Close Browser
CREATING NEW REFERENCE FILE
Browser Console Log 0: JSHandle:JQMIGRATE: Migrate is installed with logging active, version 3.4.1
Browser Console Log 0: JSHandle:BackstopTools have been installed.
x Close Browser
x Close Browser

Run `$ backstop test` to generate diff report.

      COMMAND | Command "reference" successfully executed in [28.301s]
6/14 steps [======================================================================================                                                                                                                   ] 1
1) Scenario: Beacon captures expected images in desktop # src/features/lcp-beacon-script.feature:10
   ✔ Before # src/support/hooks.ts:132
   ✔ Given I am logged in # src/support/steps/general.ts:22
   ✔ And plugin is installed 'new_release' # src/support/steps/general.ts:28
   ✔ And plugin is activated # src/support/steps/general.ts:47
   ✔ And I go to 'wp-admin/options-general.php?page=wprocket#dashboard' # src/support/steps/general.ts:85
   ✔ When I log out # src/support/steps/general.ts:128
   ✖ And I visit the urls for 'desktop' # src/support/steps/lcp-beacon-script.ts:17
       page.waitForFunction: Timeout 30000ms exceeded.
           at World.<anonymous> (/Users/mathieu/Documents/Github/wp-rocket-e2e/src/support/steps/lcp-beacon-script.ts:38:29)
   - Then lcp and atf should be as expected for 'desktop' # src/support/steps/lcp-beacon-script.ts:69
   ✔ After # src/support/hooks.ts:146
       Attachment (image/png)
       Attachment (video/webm)


1
2) Scenario: Beacon captures expected images in mobile # src/features/lcp-beacon-script.feature:15
   ✔ Before # src/support/hooks.ts:132
   ✔ Given I am logged in # src/support/steps/general.ts:22
   ✔ And plugin is installed 'new_release' # src/support/steps/general.ts:28
   ✔ And plugin is activated # src/support/steps/general.ts:47
   ✔ And I go to 'wp-admin/options-general.php?page=wprocket#dashboard' # src/support/steps/general.ts:85
   ✔ When I log out # src/support/steps/general.ts:128
   ✖ And I visit the urls for 'mobile' # src/support/steps/lcp-beacon-script.ts:17
       page.waitForFunction: Timeout 30000ms exceeded.
           at World.<anonymous> (/Users/mathieu/Documents/Github/wp-rocket-e2e/src/support/steps/lcp-beacon-script.ts:38:29)
   - Then lcp and atf should be as expected for 'mobile' # src/support/steps/lcp-beacon-script.ts:69
   ✔ After # src/support/hooks.ts:146
       Attachment (image/png)
       Attachment (video/webm)


14/14 steps [========================================================================================================================================================================================================] 2 scenarios (2 failed)
14 steps (2 failed, 2 skipped, 10 passed)
2m41.240s (executing steps: 2m04.198s)
```

Khadreal committed:
Closes #70 Add scenario for pages without LCP/ATF (#76)

* Add scenario for no lcp/atf, add plugin deactivation function #70 :closes:

* fixed lint

* minor fix

* Add step to activate and deactivate mobile plugin

* update gitignore file

* PR modifications

* Add new command function, pr modifications --#70

* minor modifications

* Add beacon check for lcp pages

* deactivate plugin, modify json

* Add missing step

* change steps

jeawhanlee committed:
Closes #82: LCP test not failing with no result in db (#84)

* Make tests fail when no result in db

* Fail tests early

* Make log red

jeawhanlee committed:
Merge branch 'trunk' into enhancement/71-use-result-file-for-lcp-tests

jeawhanlee committed:
Merge branch 'trunk' of https://github.com/wp-media/wp-rocket-e2e into trunk

jeawhanlee committed:
Merge branch 'trunk' into enhancement/71-use-result-file-for-lcp-tests

jeawhanlee commented:
@Mai-Saad I think we can merge this PR since the failing mobile scenario is to be fixed [here](https://github.com/wp-media/wp-rocket-e2e/issues/83)

Miraeld commented:
@jeawhanlee  Should I merge it ?
